File: frontend/src/utils/result.ts
export class Result<T> {
  private constructor(
    public readonly success: boolean,
    public readonly data: T | null,
    public readonly error: string | null,
    public readonly errors: Record<string, string[]> | null
  ) {}

  static ok<T>(data: T): Result<T> {
    return new Result<T>(true, data, null, null);
  }

  static fail<T = never>(error: string): Result<T> {
    return new Result<T>(false, null, error, null);
  }

  static failWithErrors<T = never>(errors: Record<string, string[]>): Result<T> {
    return new Result<T>(false, null, null, errors);
  }

  isSuccess(): this is Result<T> & { data: T } {
    return this.success;
  }

  isFailure(): this is Result<T> & { error: string } | Result<T> & { errors: Record<string, string[]> } {
    return !this.success;
  }
}

File: backend/NutriPlan.Application/Common/Result.cs
namespace NutriPlan.Application.Common;

public class Result<T>
{
    public bool Success { get; private set; }
    public T? Data { get; private set; }
    public string? Error { get; private set; }
    public Dictionary<string, string[]>? Errors { get; private set; }

    private Result(bool success, T? data, string? error, Dictionary<string, string[]>? errors = null)
    {
        Success = success;
        Data = data;
        Error = error;
        Errors = errors;
    }

    public static Result<T> Ok(T data) => new(true, data, null);
    public static Result<T> Fail(string error) => new(false, default, error);
    public static Result<T> Fail(Dictionary<string, string[]> errors) => new(false, default, null, errors);
}